test/version_solver_test.dart

```diff
@@ -1661,6 +1661,84 @@ void prerelease() {
     await d.appDir({'a': '<=2.0.0-dev'}).create();
     await expectResolves(result: {'a': '1.1.0'});
   });
+
+  test('use pre-release when desired', () async {
+    await servePackages((builder) {
+      builder.serve('a', '1.0.0');
+      builder.serve('a', '1.1.0-dev');
+    });
+
+    await d.appDir({'a': '^1.1.0-dev'}).create();
+    await expectResolves(result: {'a': '1.1.0-dev'});
+  });
+
+  test('can upgrade from pre-release', () async {
+    await servePackages((builder) {
+      builder.serve('a', '1.0.0');
+      builder.serve('a', '1.1.0-dev');
+      builder.serve('a', '1.1.0');
+    });
+
+    await d.appDir({'a': '^1.1.0-dev'}).create();
+    await expectResolves(result: {'a': '1.1.0'});
+  });
+
+  test('will use pre-release if depended on in stable release', () async {
+    // This behavior is desired because a stable package has dependency on a
+    // pre-release.
+    await servePackages((builder) {
+      builder.serve('a', '1.0.0', deps: {'b': '^1.0.0'});
+      builder.serve('a', '1.1.0', deps: {'b': '^1.1.0-dev'});
+      builder.serve('b', '1.0.0');
+      builder.serve('b', '1.1.0-dev');
+    });
+
+    await d.appDir({'a': '^1.0.0'}).create();
+    await expectResolves(result: {
+      'a': '1.1.0',
+      'b': '1.1.0-dev',
+    });
+  });
+
+  test('back-tracks pre-release choice with direct-dependency', () async {
+    await servePackages((builder) {
+      builder.serve('a', '1.0.0', deps: {'b': '^1.0.0'});
+      builder.serve('a', '1.1.0', deps: {'b': '^1.1.0-dev'});
+      builder.serve('b', '1.0.0');
+      builder.serve('b', '1.1.0-dev');
+    });
+
+    await d.appDir({
+      'a': '^1.0.0',
+      'b': '^1.0.0', // Direct-dependency prevents us from using a pre-release.
+    }).create();
+    await expectResolves(result: {
+      'a': '1.0.0',
+      'b': '1.0.0',
+    });
+  });
+
+  test('back-tracking pre-release fails with indirect-dependency', () async {
+    // NOTE: This behavior is not necessarily desired. It is a performance
+    //       compromise, as back-tracking is expensive.
```

Comment:
Is this actually a deliberate perf choice, or is it just that we didn't realize the solver would behave this way?

Reply:
I heard @jakemac53 suggest it might be a performance choice.

It could also be unintended behavior.

I haven't looked at how complicated it would be to change the behavior. Regardless, it's probably fine to retrain the behavior for now, document it with a comment in the test case and then we can always review in the future. One would hope that being accidentally bumped to a pre-release rarely happens, and rarely has negative side-effects.

Reply:
Yeah, I'm OK with the behavior as it is. I just don't want to leave a comment that claims a rationale unless we're sure that *is* the reason. Otherwise, it will just mislead future maintainers.

Reply:
Ya, that this was a performance choice is just an assumption. It would certainly be more expensive to support backtracking and ranking previous possible results or something, but I don't know any actual measurements of the cost.